ipywidgets/widgets/trait_types/__init__.py

```diff
@@ -0,0 +1,185 @@
+    """Serializes a Python datetime object to json.
+
+    Instantiating a JavaScript Date object with a string assumes that the
+    string is a UTC string, while instantiating it with constructor arguments
+    assumes that it's in local time:
+
+    >>> cdate = new Date('2015-05-12')
+    Mon May 11 2015 20:00:00 GMT-0400 (Eastern Daylight Time)
+    >>> cdate = new Date(2015, 4, 12) // Months are 0-based indices in JS
+    Tue May 12 2015 00:00:00 GMT-0400 (Eastern Daylight Time)
+
+    Attributes of this dictionary are to be passed to the JavaScript Date
+    constructor.
+    """
+    if pydt is None:
+        return None
+    else:
+        return dict(
+            year=pydt.year,
+            month=pydt.month - 1,  # Months are 0-based indices in JS
+            date=pydt.day,
+            hours=pydt.hour,       # Hours, Minutes, Seconds and Milliseconds are
+            minutes=pydt.minute,   # plural in JS
+            seconds=pydt.second,
+            milliseconds=pydt.microsecond / 1000
+        )
+
+def datetime_from_json(js, manager):
+    """Deserialize a Python datetime object from json"""
+    if js is None:
+        return None
+    else:
+        return dt.datetime(
+            js['year'],
+            js['month'] + 1,  # Months are 1-based in Python
+            js['date'],
+            js['hours'],
+            js['minutes'],
+            js['seconds'],
+            js['milliseconds'] * 1000
+        )
+
+datetime_serialization = {
+    'from_json': datetime_from_json,
+    'to_json': datetime_to_json
+}
+
+class InstanceDict(traitlets.Instance):
+    """An instance trait which coerces a dict to an instance.
+    
+    This lets the instance be specified as a dict, which is used to initialize the instance.
+    
+    Also, we default to a trivial instance, even if args and kwargs is not specified."""
+
+    def validate(self, obj, value):
+        if isinstance(value, dict):
+            return super(InstanceDict, self).validate(obj, self.klass(**value))
+        else:
+            return super(InstanceDict, self).validate(obj, value)
+
+    def make_dynamic_default(self):
+        return self.klass(*(self.default_args or ()),
+                          **(self.default_kwargs or {}))
+
+
+class EDict(eventful.Eventful, traitlets.Dict):
+
+    event_map = {
+        'setitem': ('__setitem__', 'setdefault'),
+        'delitem': ('__delitem__', 'pop'),
+        'update': 'update',
+        'clear': 'clear',
+    }
+    type_name = 'edict'
+
+    def _before_update(self, inst, call):
+        new = call.args[0]
+        new.update(call.kwargs)
+        call_list = []
+        return self.redirect(
+            None, 'setitem', inst,
+            args=new.items())
+
+    def _before_clear(self, inst, call):
+        return self.redirect(
+            None, 'delitem', inst,
+            args=inst.keys())
+
+    @staticmethod
+    def before_setitem(inst, call):
+        """Expect call.args[0] = key"""
```

Comment:
1. The description was more of a personal note. I will be writing a complete set of docstrings.
2. You're right, it'd be better to just merge `_before_setitem` and `before_setitem`.

## Why allow the return of "afterbacks"?

This is what enables redirects.

A `Redirect` is essentially a before/afterback (b/a) pair wrapped in one object:

1. `Redirect.__init__` : Gathers the `target` b/a, calls the beforeback, and creates a closure that holds the beforeback's result and will execute the afterback when called.
2. `Redirect.__call__`: triggers the aforementioned closure, which runs the afterback.

Because callable objects returned by beforebacks act like a afterbacks. A "redirection" of a notification can be performed by returning a `Redirect` from a beforeback.

Let me know if this makes sense.